src/DocBlock/TypeExpression.php

```diff
@@ -267,58 +263,78 @@ public function getTypesGlue(): string
     }
 
     /**
-     * @param \Closure(self): void $callback
+     * @param \Closure(self): self $callback
      */
-    public function walkTypes(\Closure $callback): void
+    public function mapTypes(\Closure $callback): self
     {
-        $innerValueOrig = $this->value;
-
+        $value = $this->value;
         $startIndexOffset = 0;
 
         foreach ($this->innerTypeExpressions as [
             'start_index' => $startIndexOrig,
             'expression' => $inner,
         ]) {
-            $innerLengthOrig = \strlen($inner->toString());
+            $innerValueOrig = $inner->value;
 
-            $inner->walkTypes($callback);
+            $inner = $inner->mapTypes($callback);
 
-            $this->value = substr_replace(
-                $this->value,
-                $inner->toString(),
-                $startIndexOrig + $startIndexOffset,
-                $innerLengthOrig
-            );
+            if ($inner->value !== $innerValueOrig) {
+                $value = substr_replace(
+                    $value,
+                    $inner->value,
+                    $startIndexOrig + $startIndexOffset,
+                    \strlen($innerValueOrig)
+                );
 
-            $startIndexOffset += \strlen($inner->toString()) - $innerLengthOrig;
+                $startIndexOffset += \strlen($inner->value) - \strlen($innerValueOrig);
+            }
         }
 
-        $callback($this);
+        $type = $value === $this->value
+            ? $this
+            : $this->inner($value);
 
-        if ($this->value !== $innerValueOrig) {
-            $this->isUnionType = false;
-            $this->typesGlue = '|';
-            $this->innerTypeExpressions = [];
+        return $callback($type);
+    }
 
-            $this->parse();
-        }
+    /**
+     * @param \Closure(self): void $callback
+     */
+    public function walkTypes(\Closure $callback): void
+    {
+        $type = $this->mapTypes(static function (self $type) use ($callback) {
```

Comment:
Do we need this assignment? `walkTypes` returns `void`, so it seems superfluous?

Reply:
we need is for two purposes:
- reuse `map` for `walk` impl.
- assert walk callback did not change the value (map should be used if you want to change it)

Reply:
But it's not returned and is not used in closure's context, so basically it is written but never read? Am I missing something?

Reply:
We pass object to the callback and later we assert that the value in that object in unchanged (by the called callback).

Reply:
I did not look in IDE when reviewing, but I checked it now and PHPStorm also marks this assignment as unused:

<img width="753" alt="image" src="https://github.com/user-attachments/assets/f6230f21-bfd9-42ba-85ad-5f199a05eb28">

The `$type` where the result of `mapTypes()` is assigned is _not_ the same `$type` used in closure's signature - it's not a reference, it's just a variable named the same as the one in the closure. You can remove this assignment and result of the `TypeExpressionTest` is exactly the same (BTW. I have 2 failures locally even with this assignment: `AssertionError: assert($typesGlue === $matches['glue'][0])`)

Reply:
Thank you for spotting it, I was blind here, fixed!